component/kafka/simple/consumer.go

```diff
@@ -0,0 +1,301 @@
+	if c.component.notificationOnceReachingLatestOffset() {
+		// We don't want to wait for consuming a message if we already know we're at the end of the stream
+		if c.startingOffsetAfterLatestOffset(c.component.latestOffsets[partition], partition) {
+			c.mu.Lock()
+			c.partitionHasReachedLatestOffset(partition)
+			c.mu.Unlock()
+		}
+	}
+
+	if c.isBatch {
+		return c.consumePartitionBatch(ctx, ch, partition)
+	}
+
+	return c.consumePartitionUnit(ctx, ch, partition)
+}
+
+func (c *consumerHandler) startingOffsetAfterLatestOffset(latestOffset int64, partition int32) bool {
+	return c.component.startingOffsets[partition] >= latestOffset
+}
+
+func (c *consumerHandler) partitionHasReachedLatestOffset(partition int32) {
+	delete(c.partitionsWithLatestOffsetUnreached, partition)
+	if len(c.partitionsWithLatestOffsetUnreached) == 0 {
+		c.notificationAlreadySent = true
+		// As the consumer can be retried, we have to make sure the channel is closed only once.
+		c.component.once.Do(func() {
+			close(c.component.latestOffsetReachedChan)
+		})
+	}
+}
+
+func (c *consumerHandler) consumePartitionUnit(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) error {
+	for {
+		select {
+		case msg, ok := <-ch:
+			if !ok {
+				log.Debug("messages channel closed")
+				return nil
+			}
+			log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+			messageStatusCountInc(messageReceived, partition, msg.Topic)
+			err := c.unit(ctx, msg)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if errors.Is(ctx.Err(), context.Canceled) {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) consumePartitionBatch(ctx context.Context, ch <-chan *sarama.ConsumerMessage, partition int32) (err error) {
+	// We need to enforce flushing the buffer before to return (e.g., context cancelled but messages are still buffered)
+	defer func() {
+		errFlush := c.flush(ctx)
+		if errFlush != nil {
+			err = errFlush
+		}
+	}()
+
+	for {
+		select {
+		case msg, ok := <-ch:
+			if ok {
+				log.Debugf("message claimed: value = %s, timestamp = %v, partition = %d topic = %s", string(msg.Value), msg.Timestamp, msg.Partition, msg.Topic)
+				messageStatusCountInc(messageReceived, partition, msg.Topic)
+				err := c.insertMessage(ctx, msg)
+				if err != nil {
+					return err
+				}
+			} else {
+				log.Debug("messages channel closed")
+				return nil
+			}
+		case <-c.ticker.C:
+			err := c.flush(ctx)
+			if err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			if errors.Is(ctx.Err(), context.Canceled) {
+				log.Infof("closing consumer: %v", ctx)
+			}
+			return nil
+		}
+	}
+}
+
+func (c *consumerHandler) unit(ctx context.Context, msg *sarama.ConsumerMessage) error {
+	messageStatusCountInc(messageProcessed, msg.Partition, msg.Topic)
```

Comment:
If I get this right, at this point the processing of the message is not processed, right? 
I'd expect this counter to be increased after the processing is finished.
Also, since `unit()` is called directly after the `messageReceived`, we'd have identical counters for these two stages for non-batch processing